documentation/specs/static-graph.md

```diff
@@ -143,49 +139,100 @@ This is a proposal for what the public API may look like:
+                 v    v
+                 Proj4   A->B, A
+```
+
+The common project reference protocols (Build, Rebuild, Restore, Clean) will be specified by the common props and targets file in the msbuild repository. Other SDKs can implement their own protocols (e.g. ASPNET implementing Publish).
+
+Here are the rules for the common protocols:
+
+$Build \to GetTargetFrameworks, \langle default \rangle, GetNativeManifest, GetCopyToOutputDirectoryItems$
+
+The default target (represented here as $\langle default \rangle$) is resolved for each project.
+
+$Clean \to GetTargetFrameworks, Clean$
+
+$Rebuild \to \$Clean, \$Build$
+
+$Rebuild$ maps to two referenced protocols. $\$Clean$ and $\$Build$ get substituted with the contents of the $Clean$ and $Build$ protocols.
+
+Restore is a composition of two rules:
+- $Restore \to \_IsProjectRestoreSupported, \_GenerateRestoreProjectPathWalk, \_GenerateRestoreGraphProjectEntry$
 
-This makes the P2P contract explicit as it requires both projects to declare targets which are part of the contract.
+- $\_GenerateRestoreProjectPathWalk \to \_IsProjectRestoreSupported, \_GenerateRestoreProjectPathWalk, \_GenerateRestoreGraphProjectEntry$
+
+**Open Issue:** Restore is a bit complicated, and we may need new concepts to represent it. The root project calls the recursive $\_GenerateRestoreProjectPathWalk$ on itself to collect the children closure, and then after the recursion call returns (after having walked the graph), it calls the other targets on each returned child in a non-recursive manner. So the above protocol is not a truthful representation of what happens, but it correctly captures all targets called on each node in the graph.
+
+**Open Issue:** How do we differentiate between targets called on the outer build project and targets called on the inner build projects?
+
+We'll represent the project reference protocols as items in MSBuild:
+```xml
+<PropertyGroup>
+  <ProjectReferenceTargetsForClean>GetTargetFrameworks;Clean</ProjectReferenceTargetsForClean>
+</PropertyGroup>
+
+<ItemGroup>
+  <ProjectReferenceTargets Include="Clean" Targets="$(ProjectReferenceTargetsForClean)"/>
+</ItemGroup>
+```
+
+## Building a single project
+The project graph is required to build the projects in the correct order. An individual project will be built using `/p:BuildProjectReferences=false` to avoid recursive builds inside a project.
+
+**OPEN ISSUE:** Can we get away with not setting `/p:BuildProjectReferences=false`?
+
+When building a project, run the initial targets and the entry targets as computed by the [project reference protocol](#inferring-which-targets-to-run-for-a-project-within-the-graph).
+
+After execution, the results of the entry targets will be cached in the BuildManager and the project can be disposed. Note that this gives project execution a concrete start and end, while previously projects needed to remain active just in case any other project needed to execute a target on it.
```

Comment:
Ah, I didn't think about initial targets. It seems like we would not want to "export" those, since they're not part of the p2p protocol.

Reply:
Correct. I'll call that out.